### PR TITLE
Document new file permission API

### PR DESCRIPTION
**Pull request by jbartok**
Fixes
* #24857

### CODE REVIEW

**Comment by eskatos** [comment]
@bot-gradle test BD

File: subprojects/docs/src/snippets/files/copy/groovy/build.gradle [hunk 1]
```diff
@@ -234,6 +234,23 @@ tasks.register('filter', Copy) {
 }
 // end::filter-files[]
 
+// tag::file-permissions[]
+tasks.register('rename', Copy) {
```
Comment (eskatos) [inline subprojects/docs/src/snippets/files/copy/groovy/build.gradle:238]:
```suggestion
tasks.register('permissions', Copy) {
```

File: subprojects/core-api/src/main/java/org/gradle/api/file/FileTreeElement.java [hunk 2]
```diff
@@ -108,5 +108,5 @@ public interface FileTreeElement {
      * @since 8.3
      */
     @Incubating
-    FilePermissions getImmutablePermissions();
+    FilePermissions getPermissions();
```
Comment (eskatos) [inline subprojects/core-api/src/main/java/org/gradle/api/file/FileTreeElement.java:111]: :+1:

File: subprojects/docs/src/docs/userguide/authoring-builds/working_with_files.adoc [hunk 3]
```diff
@@ -642,6 +642,30 @@ The `expand()` method treats the source files as https://docs.groovy-lang.org/la
 
 NOTE: It's good practice to specify the character set when reading and writing the file, otherwise the transformations won't work properly for non-ASCII text. You configure the character set with the link:{javadocPath}/org/gradle/api/file/CopySpec.html#setFilteringCharset-java.lang.String-[CopySpec.setFilteringCharset(String)] property. If it's not specified, the JVM default character set is used, which is likely to be different from the one you want.
 
+[[sec:setting_file_permissions]]
+=== Setting file permissions
+
+For any `CopySpec` involved in copying files, may it be the `Copy` task itself, or any child specifications, you can explicitly set the permissions the destination files will have, via the link:{javadocPath}/org/gradle/api/file/CopySpec.html#filePermissions-org.gradle.api.Action-[CopySpec.filePermissions {}] configurations block.
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/authoring-builds/working_with_files.adoc:648]: The javadoc link in this paragraph currently points to the top of the `CopySpec`  page. The suggestion below makes it point directly to the `filePermissions {}` method but in the `CopyProcessingSpec` page. I think this is better, your call.

```suggestion
For any `CopySpec` involved in copying files, may it be the `Copy` task itself, or any child specifications, you can explicitly set the permissions the destination files will have, via the link:{javadocPath}/org/gradle/api/file/CopyProcessingSpec.html#filePermissions-org.gradle.api.Action-[CopySpec.filePermissions {}] configurations block.
```

File: subprojects/docs/src/docs/userguide/authoring-builds/working_with_files.adoc [hunk 4]
```diff
@@ -642,6 +642,30 @@ The `expand()` method treats the source files as https://docs.groovy-lang.org/la
...
+You can do the same for directories too, independently of files, via the link:{javadocPath}/org/gradle/api/file/CopySpec.html#dirPermissions-org.gradle.api.Action-[CopySpec.dirPermissions {}] configurations block.
```
Comment (eskatos) [inline subprojects/docs/src/docs/userguide/authoring-builds/working_with_files.adoc:649]: Same as above

```suggestion
You can do the same for directories too, independently of files, via the link:{javadocPath}/org/gradle/api/file/CopyProcessingSpec.html#dirPermissions-org.gradle.api.Action-[CopySpec.dirPermissions {}] configurations block.
```

**Review by eskatos — APPROVED** [review]
:+1:

**Comment by jbartok** [comment]
@bot-gradle test this

**Comment by jbartok** [comment]
@bot-gradle test and merge

**Pull request merged**